### PR TITLE
Sanitize Kafka Connect config secrets

### LINKED ISSUE #13: Do not display secrets in Kafka connect configs

### PR DESCRIPTION
**Pull request by disrupted**
Closes #13

### CODE REVIEW

File: backend/streams_explorer/core/services/kafkaconnect.py [hunk 1]
```diff
@@ -18,11 +18,45 @@ def get_connectors() -> list:
...
+        response = requests.put(
+            f"{url}/connector-plugins/{connector_class}/config/validate",
+            json=config,
+        )
+        if not response.ok:
+            return config
+        data = response.json()
+        protected_keys = [
+            config["value"]["name"]
+            for config in data["configs"]
```
Comment (philipp94831) [inline backend/streams_explorer/core/services/kafkaconnect.py:52]: just use definition.type == PASSWORD

**Review by philipp94831 — CHANGES_REQUESTED** [review]
Please add a test

File: backend/streams_explorer/core/services/kafkaconnect.py (hunk 1 above)
Comment (philipp94831) [reply backend/streams_explorer/core/services/kafkaconnect.py:52]:
```json
"definition":{
        "name":"connection.password",
        "type":"PASSWORD",
        "required":false,
        "default_value":null,
        "importance":"MEDIUM",
        "documentation":"The password used to authenticate with Elasticsearch. The default is the null, and authentication will only be performed if  both the username and password are non-null.",
        "group":"Connector",
        "width":"SHORT",
        "display_name":"Connection Password",
        "dependents":[
        ],
        "order":3
      },
```

File: backend/streams_explorer/core/services/kafkaconnect.py [hunk 2]
```diff
@@ -18,11 +18,45 @@ def get_connectors() -> list:
...
+            return connector_class.rsplit(".", 1)[-1]
+        return connector_class
+
+    @staticmethod
+    @logger.catch
+    def sanitize_connector_config(config: dict) -> dict:
+        connector_class = KafkaConnect.extract_connector_class_basename(
+            config["connector.class"]
+        )
+        response = requests.put(
```
Comment (philipp94831) [inline backend/streams_explorer/core/services/kafkaconnect.py:43]: What about caching the config result per connector class? The schema is the same for them

File: backend/streams_explorer/core/services/kafkaconnect.py (hunk 1 above)
Comment (disrupted) [reply backend/streams_explorer/core/services/kafkaconnect.py:52]: I think there is no point in hiding fields which are not used.

Comment (philipp94831) [reply backend/streams_explorer/core/services/kafkaconnect.py:52]: I just think that `[hidden]` is not reliable but PASSWORD is part of the API definition. Also that way you just collect the fields that contain passwords and then only apply it to the fields of the config

File: backend/streams_explorer/core/services/kafkaconnect.py (hunk 2 above)
Comment (disrupted) [reply backend/streams_explorer/core/services/kafkaconnect.py:43]: Good idea, however I don't know what's the preferred way of caching with `requests`. I tried it with [`requests-cache`](https://pypi.org/project/requests-cache/) but can't get caching to work for only this endpoint, without monkeypatching the whole requests module and then having to disable it for the other uses.

Another approach I tried was using
```python
cached_session = requests_cache.CachedSession()

def foo(self):
    response = cached_session.put(...)
```

but `response.from_cache` was always `False`

Comment (philipp94831) [reply backend/streams_explorer/core/services/kafkaconnect.py:43]: No need to make it that complicated. Just store the results in a Dict :)

Comment (philipp94831) [reply backend/streams_explorer/core/services/kafkaconnect.py:43]: Also you can just store the list of password fields, no need to store the whole config

Comment (disrupted) [reply backend/streams_explorer/core/services/kafkaconnect.py:43]: Alright, yeah this should work.

Comment (disrupted) [reply backend/streams_explorer/core/services/kafkaconnect.py]: Tests & changes are added now :) @philipp94831

File: backend/streams_explorer/core/services/kafkaconnect.py [hunk 3]
```diff
@@ -18,11 +19,45 @@ def get_connectors() -> list:
...
+            return connector_class.rsplit(".", 1)[-1]
+        return connector_class
+
+    @staticmethod
+    def sanitize_connector_config(config: dict) -> dict:
+        connector_class = KafkaConnect.extract_connector_class_basename(
+            config["connector.class"]
+        )
+        if connector_class not in protected_keys:
+            response = requests.put(
```
Comment (philipp94831) [inline backend/streams_explorer/core/services/kafkaconnect.py:44]: extract a method

File: backend/streams_explorer/core/services/kafkaconnect.py [hunk 4]
```diff
@@ -8,6 +8,7 @@
 from streams_explorer.models.kafka_connector import KafkaConnector
 
 url = settings.kafkaconnect.url
+protected_keys: Dict[str, Set[str]] = {}
```
Comment (philipp94831) [inline backend/streams_explorer/core/services/kafkaconnect.py:11]: Should we clear this upon every connector refresh?

**Review by philipp94831 — CHANGES_REQUESTED** [review]

File: backend/streams_explorer/core/services/kafkaconnect.py (hunk 4 above)
Comment (disrupted) [reply backend/streams_explorer/core/services/kafkaconnect.py:11]: do you mean the `= {}` part?

Comment (philipp94831) [reply backend/streams_explorer/core/services/kafkaconnect.py:11]: Yeah, it get's filled but never emptied, right? What if we update the connectors while streams-explorer is running?

Comment (disrupted) [reply backend/streams_explorer/core/services/kafkaconnect.py:11]: Hm, the Connector class definition doesn't change, right? If we add or remove Connector classes this wouldn't have to be cleared unless there's a specific reason imo. Or do you mean something different?

Comment (philipp94831) [reply backend/streams_explorer/core/services/kafkaconnect.py:11]: The connector class can be updated :-)

File: backend/streams_explorer/core/services/kafkaconnect.py (hunk 1 above)
Comment (disrupted) [reply backend/streams_explorer/core/services/kafkaconnect.py:56]: Changed it, with the caching in place now it makes perfect sense.

File: backend/streams_explorer/core/services/kafkaconnect.py (hunk 3 above)
Comment (disrupted) [reply backend/streams_explorer/core/services/kafkaconnect.py]: Done.

**Review by philipp94831 — APPROVED** [review]

**Pull request merged**